### ISSUE TITLE
Warning: Skipping line with longer outputs than inputs

### ISSUE THREAD

**Issue by abhikatoldtrafford** [Calamari-OCR/calamari#206]
Hi @ChWick @andbue while I was training calamari-ocr, I am consistently getting: 'Warning: Skipping line with longer outputs than inputs' for larger images having more width. Is there any way to fix this problem? I want them to be included in the training dataset, not just supress the warning. Thanks

**Comment by andbue** [comment]
The problem is not the size of the image, but the size of the text. The network can at most handle as many characters per line as there are horizontal pixels in the image (+1 pixel for every repeated letter).
Could it be the case that some of your line images are rotated by 90 degrees? This would explain why your images are too short for your texts.
Another possibility could be that you have set a --text-normalization that breaks up diacritics or other signs that are written in combination with each other, resulting in a larger codec and too many chars per line for the network.

**Comment by ChWick** [comment]
@abhikatoldtrafford I would also assume (but please verify this) that the lines that raise this warning are corrupt or not usable for training. If you are uncertain, please share some of your lines

**Comment by abhikatoldtrafford** [comment]
Hi @ChWick @andbue  my images are not rotated by 90 degree. Just check one example:
[joined_9_th_dec_5870_tilda_23_th_nov_701_23_th_nov_97.gt.txt](https://github.com/Calamari-OCR/calamari/files/5971526/joined_9_th_dec_5870_tilda_23_th_nov_701_23_th_nov_97.gt.txt)
![joined_9_th_dec_5870_tilda_23_th_nov_701_23_th_nov_97](https://user-images.githubusercontent.com/11613662/107764737-b5d23780-6d56-11eb-9a1e-5fe2c9072e24.png)

**Comment by ChWick** [comment]
This line seems fine. It this one of the lines that are skipped?

**Comment by abhikatoldtrafford** [comment]
@ChWick yes this is one of the images skipped. I copied the path from the warning and downloaded this image. I have lots of similar images which are getting skipped.

**Comment by andbue** [comment]
I think the inverted part is a problem here. The center normalizer fails to find a center line and ends up padding large amounts of white at the top and at the bottom. After scaling to 48px, it ends up at less than 350px width.
Commenting out the dewarping part of the CenterNormalizer would avoid this. The cleanest way to handle this kind of line would probably be to apply some preprocessing that transforms partially inverted text correctly.

**Comment by abhikatoldtrafford** [comment]
@andbue inverted image is part of the augmentation I use. I have lots of images which I add to create these kind of images. You can get that from the filename.

**Comment by abhikatoldtrafford** [comment]
@andbue where should I comment the dewrapping part out?

**Comment by ChWick** [comment]
This was my current guess, too. Thanks for verifying this @andbue . The preproc fails. For now you could simply try to skip the dewarper (since your lines are flat anyways): 
add the `--data_preprocessing DataRangeNormalizer ScaleToHeightProcessor FinalPreparation` parameter, or disable inverting as augmentation

**Comment by abhikatoldtrafford** [comment]
Should I do calamari-train...--data_preprocessing DataRangeNormalizer ScaleToHeightProcessor FinalPreparation...

**Comment by ChWick** [comment]
Correct

**Comment by ChWick** [comment]
@abhikatoldtrafford  Is it working?

**Comment by abhikatoldtrafford** [comment]
I will update you tomorrow as it has just started running now. It takes 4 hours to complete an iteration.

**Comment by abhikatoldtrafford** [comment]
@ChWick I get this error: 'DataParams' object has no attribute 'line_height'. Which was previously not present...

**Comment by ChWick** [comment]
Thanks for pointing out this bug!

**Comment by abhikatoldtrafford** [comment]
@ChWick @andbue I tried center normalization using the code from 
Calamari-OCR/calamari/blob/master/calamari_ocr/ocr/dataset/imageprocessors/center_normalizer.py
for the image. Here is what I found:

![dewarped](https://user-images.githubusercontent.com/11613662/108811294-fccff080-75d2-11eb-86fa-a0108a3a6fad.jpg)
![normalized_img](https://user-images.githubusercontent.com/11613662/108811296-fe011d80-75d2-11eb-8a06-026476c3eb32.jpg)
So, nothing wrong with these images. Why do I get the warnings then? Could you look into this?

**Comment by ChWick** [comment]
@abhikatoldtrafford The CenterNormalizer is relying on white background and black foreground. It is hard to redesign the processing to also deal with mixed lines. Therefore, you have to stick with the ScaleToHeight processer for your case, or drop these kinds of augmentations. But based on your lines, the scale to height processor should work just fine, since dewarping should not be required.

**Comment by andbue** [comment]
You could also try to set the first value of the "extra_parameters" argument in CenterNormalizer to some smaller value than the default 4. When there are heavily warped lines, this could lead to some parts being cut off, but in your case it should be fine.

I'm not sure, however, that the partly inverted images do not impair the overall performance of the model. If you have any results on that, it would be interesting if you could share your evalution results with and without inverting with us here!

**Comment by abhikatoldtrafford** [comment]
@andbue thanks, value of 2 does fix the issue a bit. However, I dont know why it should degrade the model performance. Is there any particular reason behind your intuition?

**Comment by andbue** [comment]
I guess the model would have to either learn two versions for every character (normal and inverted) or learn to invert characters or come up with some efficient kind of edge detection. When your data includes a lot of lines that look like this the question would be if it produces more errors to force the OCR model to cope with these problems or to split and invert the lines before throwing them into calamari. But I'm just guessing here, I'm really interested in your empirical results!

**Comment by abhikatoldtrafford** [comment]
@andbue @ChWick I have changed "extra_parameters" argument in CenterNormalizer from 4 to 2. I have restarted training with this value. Now, during prediction, do I need to change anything? Does CenterNormalizer get called in prediction time as well?

**Comment by andbue** [comment]
If your prediction dataset contains inverted lines, you should also use the modified CenterNormalizer for that. Otherwise your special lines would be skipped here as well.

The option --data_preprocessing currently does not exist in the predict script. The model automatically loads the same preprocessors that have been used for training.
